### PR TITLE
[fx] refactor code for profiler / enable fake tensor movement.

### PR DESCRIPTION
**Pull request by super-dainiu**
### What's new?
I modified the `Tensor.to` behavior so that moving tensor from and to `fake_devices` will not be misleading.
```python
>>> tensor = MetaTensor(torch.rand(10), fake_device='cuda:100')
>>> tensor.to(torch.uint8)
MetaTensor(tensor(..., device='meta', size=(10,), dtype=torch.uint8), fake_device='cuda:100')
>>> tensor.to(torch.device('cuda:42'))
MetaTensor(tensor(..., device='meta', size=(10,)), fake_device='cuda:42')
>>> tensor.to('vulkan')
MetaTensor(tensor(..., device='meta', size=(10,)), fake_device='vulkan')
```
I also removed some redundant or duplicated code.
### Test
image

### CODE REVIEW

File: colossalai/fx/profiler/tensor.py [hunk 1]
```diff
@@ -74,3 +83,32 @@ def wrap(x):
...
+
+    @to.register
+    def _(self, device: str, dtype: Optional[_dtype] = None, non_blocking: _bool = False, copy: _bool = False) -> torch.Tensor:
+        result = super().to(dtype, non_blocking, copy) if dtype is not None else self
+        return MetaTensor(deepcopy(result), fake_device=device)
+
+    @to.register
+    def _(self, device: _device, dtype: Optional[_dtype] = None, non_blocking: _bool = False, copy: _bool = False) -> torch.Tensor:
+        result = super().to(dtype, non_blocking, copy) if dtype is not None else self
+        return MetaTensor(deepcopy(result), fake_device=device)
```
Comment (super-dainiu) [inline colossalai/fx/profiler/tensor.py:114]: Since python has no @overload, we can only inherit `torch._C._TensorBase` methods in this way.

Comment (super-dainiu) [reply colossalai/fx/profiler/tensor.py:114]: oops I did not mean to @ this user lol

File: colossalai/fx/passes/algorithms/ckpt_solver_pofo.py [hunk 2]
```diff
@@ -507,6 +508,7 @@ def solver_pofo(gm: ColoGraphModule,
     mem_limit -= parameter_size(gm)
 
     # prepare data
+    data = MetaTensor(data, fake_device=next(gm.parameters()).device)
```
Comment (super-dainiu) [inline colossalai/fx/passes/algorithms/ckpt_solver_pofo.py:511]: After last PR update, any data that is used in MetaInfoProp should first be wrapped in `MetaTensor`

File: colossalai/fx/profiler/tensor.py (hunk 1 above)
Comment (Cypher30) [reply colossalai/fx/profiler/tensor.py:114]: Use `@overload` in comment instead lol

File: colossalai/fx/passes/algorithms/ckpt_solver_pofo.py [hunk 3]
```diff
@@ -507,6 +507,10 @@ def solver_pofo(gm: ColoGraphModule,
     mem_limit -= parameter_size(gm)
 
     # prepare data
+    from colossalai import META_COMPATIBILITY
+    if META_COMPATIBILITY:
+        from colossalai.fx.profiler import MetaTensor
```
Comment (Cypher30) [inline colossalai/fx/passes/algorithms/ckpt_solver_pofo.py:512]: Import it on the top of the file lol

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py [hunk 4]
```diff
@@ -340,7 +339,10 @@ def solver_rotor(gm: ColoGraphModule,
 
     node_list = linearize(gm, cnode)
     mem_unit = mem_limit * (1.0 - eps) // mem_slots
-    data = MetaTensor(data, fake_device=next(gm.parameters()).device)
+    from colossalai import META_COMPATIBILITY
```
Comment (Cypher30) [inline colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:342]: This one is the same

File: colossalai/fx/passes/algorithms/ckpt_solver_pofo.py (hunk 3 above)
Comment (super-dainiu) [reply colossalai/fx/passes/algorithms/ckpt_solver_pofo.py:512]: This is for compatibility.

Comment (super-dainiu) [reply colossalai/fx/passes/algorithms/ckpt_solver_pofo.py:512]: torch 11 will do `MetaInfoProp` with a normal tensor

File: colossalai/fx/profiler/tensor.py (hunk 1 above)
Comment (Cypher30) [inline colossalai/fx/profiler/tensor.py:114]: These two functions are the same?

Comment (Cypher30) [reply colossalai/fx/profiler/tensor.py:114]: okay I got your idea

**Review by Cypher30 — APPROVED** [review]
Great Work!

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**